artichoke-backend/src/extn/core/enumerable/enumerable.rb

```diff
@@ -0,0 +1,450 @@
+        i -= 1
+        break if i.zero?
+      end
+      ary
+    else
+      raise ArgumentError, "wrong number of arguments (given #{args.length}, expected 0..1)"
+    end
+  end
+
+  def count(elem=NONE, &block)
+    count = 0
+    if block
+      self.each do |*val|
+        count += 1 if block.call(*val)
+      end
+    elsif elem == NONE
+      self.each { count += 1 }
+    else
+      self.each do |*val|
+        count += 1 if val.__svalue == elem
+      end
+    end
+    count
+  end
+
+  def flat_map(&block)
+    return to_enum :flat_map unless block
+
+    ary = []
+    self.each do |*e|
+      e2 = block.call(*e)
+      if e2.respond_to? :each
+        e2.each {|e3| ary.push(e3) }
+      else
+        ary.push(e2)
+      end
+    end
+    ary
+  end
+  alias collect_concat flat_map
+
+  def max_by(&block)
+    return to_enum :max_by unless block
+
+    first = true
+    max = nil
+    max_cmp = nil
+
+    self.each do |*val|
+      if first
+        max = val.__svalue
+        max_cmp = block.call(*val)
+        first = false
+      elsif (cmp = block.call(*val)) > max_cmp
+        max = val.__svalue
+        max_cmp = cmp
+      end
+    end
+    max
+  end
+
+  def min_by(&block)
+    return to_enum :min_by unless block
+
+    first = true
+    min = nil
+    min_cmp = nil
+
+    self.each do |*val|
+      if first
+        min = val.__svalue
+        min_cmp = block.call(*val)
+        first = false
+      elsif (cmp = block.call(*val)) < min_cmp
+        min = val.__svalue
+        min_cmp = cmp
+      end
+    end
+    min
+  end
+
+  def minmax(&block)
+    max = nil
+    min = nil
+    first = true
+
+    self.each do |*val|
+      val = val.__svalue
+      if first
+        max = val
+        min = val
+        first = false
+      elsif block
+        max = val if block.call(val, max).positive?
+        min = val if block.call(val, min).negative?
+      else
+        max = val if (val <=> max).positive?
+        min = val if (val <=> min).negative?
+      end
+    end
+    [min, max]
+  end
+
+  def minmax_by(&block)
+    return to_enum :minmax_by unless block
+
+    max = nil
+    max_cmp = nil
+    min = nil
+    min_cmp = nil
+    first = true
+
+    self.each do |*val|
+      if first
+        max = min = val.__svalue
+        max_cmp = min_cmp = block.call(*val)
+        first = false
+      elsif (cmp = block.call(*val)) > max_cmp
+          max = val.__svalue
+          max_cmp = cmp
+      elsif (cmp = block.call(*val)) < min_cmp
+          min = val.__svalue
+          min_cmp = cmp
+      end
+    end
+    [min, max]
+  end
+
+  def none?(pat=NONE, &block)
+    if pat != NONE
+      self.each do |*val|
+        return false if pat === val.__svalue # rubocop:disable Style/CaseEquality # rubocop:disable Style/CaseEquality
+      end
+    elsif block
+      self.each do |*val|
+        return false if block.call(*val)
+      end
+    else
+      self.each do |*val|
+        return false if val.__svalue
```

Comment:
Methods that begin with an underscore (`_`) are mruby internal methods and are not part of Ruby Core.

This method turns val into an array with the `do |*val|` and unpacks it with `__svalue` which returns the first item in the array.

I'm not sure why this function is implemented this way, but it seems like we can simplify the implementation to remove the dependency on `__svalue`.